### PR TITLE
SourceTrackers Mk2: Actually working edition

### PR DESCRIPTION
**Pull request by miratepuffin**

### CODE REVIEW

**Comment by miratepuffin** [comment]
Have now added a different fix for the client to block until its own messages have been sent. Have tested with sending two sets of updates and queries in-between

image

```
15:50:00.365 [io-compute-2] INFO  com.raphtory.internals.components.partition.PartitionOrchestrator$ - Creating '1' Partition Managers for 'conventional_baby-blue_muskox'.
15:50:01.752 [io-compute-6] INFO  com.raphtory.internals.components.partition.PartitionManager - Partition 0: Starting partition manager for 'conventional_baby-blue_muskox'.
15:50:03.026 [spawner-akka.actor.default-dispatcher-3] INFO  com.raphtory.internals.components.querymanager.QueryManager - Source '0' is unblocking analysis for Graph 'conventional_baby-blue_muskox' with 7947 messages sent.
15:50:03.167 [io-compute-7] INFO  com.raphtory.api.querytracker.QueryProgressTracker - Job ConnectedComponents_4389634494061553108: Starting query progress tracker.
15:50:03.179 [spawner-akka.actor.default-dispatcher-3] INFO  com.raphtory.internals.components.querymanager.QueryManager - Query 'ConnectedComponents_4389634494061553108' currently blocked, waiting for ingestion to complete.
15:50:03.345 [spawner-akka.actor.default-dispatcher-3] INFO  com.raphtory.internals.components.querymanager.QueryManager - Query 'ConnectedComponents_4389634494061553108' received, your job ID is 'ConnectedComponents_4389634494061553108'.
15:50:03.362 [spawner-akka.actor.default-dispatcher-3] INFO  com.raphtory.internals.components.partition.QueryExecutor - ConnectedComponents_4389634494061553108_0: Starting QueryExecutor.
15:50:03.611 [spawner-akka.actor.default-dispatcher-10] INFO  com.raphtory.internals.components.querymanager.QueryHandler - Job 'ConnectedComponents_4389634494061553108': Perspective at Time '32674' took 234 ms to run. 
15:50:03.612 [spawner-akka.actor.default-dispatcher-7] INFO  com.raphtory.api.querytracker.QueryProgressTracker - Job 'ConnectedComponents_4389634494061553108': Perspective '32674' finished in 446 ms.
15:50:03.612 [spawner-akka.actor.default-dispatcher-7] INFO  com.raphtory.api.querytracker.QueryProgressTracker - Job ConnectedComponents_4389634494061553108: Running query, processed 1 perspectives.
15:50:03.615 [spawner-akka.actor.default-dispatcher-7] INFO  com.raphtory.api.querytracker.QueryProgressTracker - Job ConnectedComponents_4389634494061553108: Query completed with 1 perspectives and finished in 449 ms.
15:50:04.951 [spawner-akka.actor.default-dispatcher-10] INFO  com.raphtory.internals.components.querymanager.QueryManager - Source '0' is unblocking analysis for Graph 'conventional_baby-blue_muskox' with 15894 messages sent.
15:50:04.960 [io-compute-4] INFO  com.raphtory.api.querytracker.QueryProgressTracker - Job Degree_5320277040710172637: Starting query progress tracker.
15:50:04.979 [spawner-akka.actor.default-dispatcher-10] INFO  com.raphtory.internals.components.querymanager.QueryManager - Query 'Degree_5320277040710172637' currently blocked, waiting for ingestion to complete.
15:50:05.160 [spawner-akka.actor.default-dispatcher-10] INFO  com.raphtory.internals.components.querymanager.QueryManager - Query 'Degree_5320277040710172637' received, your job ID is 'Degree_5320277040710172637'.
15:50:05.162 [spawner-akka.actor.default-dispatcher-10] INFO  com.raphtory.internals.components.partition.QueryExecutor - Degree_5320277040710172637_0: Starting QueryExecutor.
15:50:05.214 [spawner-akka.actor.default-dispatcher-10] INFO  com.raphtory.internals.components.querymanager.QueryHandler - Job 'Degree_5320277040710172637': Perspective at Time '32674' took 49 ms to run. 
15:50:05.214 [spawner-akka.actor.default-dispatcher-11] INFO  com.raphtory.api.querytracker.QueryProgressTracker - Job 'Degree_5320277040710172637': Perspective '32674' finished in 254 ms.
15:50:05.215 [spawner-akka.actor.default-dispatcher-11] INFO  com.raphtory.api.querytracker.QueryProgressTracker - Job Degree_5320277040710172637: Running query, processed 1 perspectives.
15:50:05.215 [spawner-akka.actor.default-dispatcher-11] INFO  com.raphtory.api.querytracker.QueryProgressTracker - Job Degree_5320277040710172637: Query completed with 1 perspectives and finished in 255 ms.
```

File: core/src/main/scala/com/raphtory/internals/management/QuerySender.scala [hunk 1]
```diff
@@ -51,6 +54,13 @@ private[raphtory] class QuerySender(
...
 
   def submit(query: Query, customJobName: String = ""): QueryProgressTracker = {
+
+    if (lastQueryIndex < individualUpdateIndex) {
+      unblockIngestion(sourceID = sourceID, individualUpdateIndex, force = false)
+      lastQueryIndex = individualUpdateIndex
+      blockingSources += sourceID
+    }
+
     val jobName     = if (customJobName.nonEmpty) customJobName else getDefaultName(query)
```
Comment (ljeub-pometry) [inline core/src/main/scala/com/raphtory/internals/management/QuerySender.scala:64]: This creates a race condition in the case of multiple sets of updates and queries. We need to integrate the unblock information into the query message itself.

**Review by ljeub-pometry — CHANGES_REQUESTED** [review]
Has a race condition for the auto blocking

File: core/src/main/scala/com/raphtory/internals/management/QuerySender.scala [hunk 2]
```diff
@@ -33,32 +33,47 @@ private[raphtory] class QuerySender(
...
+      }
+
+    currentSourceID
+  }
+
+  def getIndex: Int = totalUpdateIndex
 
   def submit(query: Query, customJobName: String = ""): QueryProgressTracker = {
 
-    if (lastQueryIndex < individualUpdateIndex) {
```
Comment (ljeub-pometry) [inline core/src/main/scala/com/raphtory/internals/management/QuerySender.scala]: could just check for `indexSinceLastIDChange > 0` and remove `lastQueryIndex` (one less thing to keep track of)

File: core/src/main/scala/com/raphtory/internals/management/QuerySender.scala [hunk 3]
```diff
@@ -33,32 +33,47 @@ private[raphtory] class QuerySender(
...
+  private var totalUpdateIndex       = 0     //used at the secondary index for the client
+  private var indexSinceLastIDChange = 0     //used to know how many messages to wait for when blocking in the Q manager
+  private var lastQueryIndex         = 0     //used to see if a prior query already sent out the unblocking message
+  private var newIDRequiredOnUpdate  = false // has a query been since since the last update and do I need a new ID
+  private var currentSourceID        = -1
+
+  def sourceID(update: Boolean): Int = {
+
+    if (
+            currentSourceID == -1 || (update && newIDRequiredOnUpdate)
```
Comment (ljeub-pometry) [inline core/src/main/scala/com/raphtory/internals/management/QuerySender.scala:53]: `newIDRequiredOnUpdate == (indexSinceLastIDChange == 0)`

could do everything using only `indexSinceLastIDChange` and `totalUpdateIndex` not sure if it is a good idea though

**Review by ljeub-pometry — APPROVED** [review]
Could simplify a little but looks like it should work

**Pull request merged**